airbyte-ci/connectors/metadata_service/lib/metadata_service/docker_hub.py

```diff
@@ -27,6 +27,75 @@ def get_docker_hub_auth_token() -> str:
     return token
 
 
+def get_docker_hub_headers() -> Dict | None:
+    if "DOCKER_HUB_USERNAME" not in os.environ or "DOCKER_HUB_PASSWORD" not in os.environ:
+        # If the Docker Hub credentials are not provided, we can only anonymously call the Docker Hub API.
+        # This will only work for public images and lead to a lower rate limit.
+        return {}
+    else:
+        token = get_docker_hub_auth_token()
+        return {"Authorization": f"JWT {token}"} if token else {}
+
+
+def get_docker_hub_tags_and_digests(
+    image_name: str,
+    retries: int = 0,
+    wait_sec: int = 30,
+    next_page_url: str | None = None,
+    tags_and_digests: Dict[str, str] | None = None,
+    paginate: bool = True,
+) -> Dict[str, str]:
+    """Find all released tags and digests for an image.
+
+    Args:
+        image_name (str): The image name to get tags and digest
+        retries (int, optional): The number of times to retry the request. Defaults to 0.
+        wait_sec (int, optional): The number of seconds to wait between retries. Defaults to 30.
+        next_page_url (str | None, optional): The next DockerHub page to consume. Defaults to None.
+        tags_and_digest (Dict[str, str] | None, optional): The accumulated tags and digests for recursion. Defaults to None.
+
+    Returns:
+        Dict[str, str]: Mapping of image tag to digest
+    """
+    headers = get_docker_hub_headers()
+    tags_and_digests = tags_and_digests or {}
+
+    if not next_page_url:
+        tags_url = f"https://registry.hub.docker.com/v2/repositories/{image_name}/tags"
+    else:
+        tags_url = next_page_url
+
+    # Allow for retries as the DockerHub API is not always reliable with returning the latest publish.
+    for _ in range(retries + 1):
+        response = requests.get(tags_url, headers=headers)
+        if response.ok:
+            break
+            # This is to handle the case when a connector has not ever been released yet.
+        if response.status_code == 404:
+            return tags_and_digests
```

Comment:
Does the comment on line 73 go with these lines? 

I feel like this could end up confusing us if we need to debug something; naively it looks like this could return a partial response, if we get some results on the first page and then a 404 on the second. I'd prefer to `raise_for_status` on 404s and let this raise an exception that we catch and log in `get_latest_version_on_dockerhub`, and never allow ourselves to return a partial response.

Reply:
> Does the comment on line 73 go with these lines?

Yes.

I added a `raise_for_status` on the for loop exit. This should raise an error when, after retry, a different error status than 404 is returned by the API.
I also printed some messages when we face a 404. (this library has no logger and uses `print` in other context...)